python/lsst/dax/ppdbx/gcp/bq.py

```diff
@@ -0,0 +1,477 @@
+        str
+            The location of the BigQuery dataset associated with this
+            QueryRunner instance. This is typically the region where the
+            dataset is hosted.
+        """
+        return self._location
+
+    @classmethod
+    def log_job(cls, job: AnyBQJob, label: str, level: int = logging.DEBUG) -> None:
+        """Log details of a BigQuery job.
+
+        Parameters
+        ----------
+        job : bigquery.job.QueryJob
+            The BigQuery job to log.
+        label : str
+            A label for the job, typically indicating the type of operation
+            (e.g., "insert", "delete", "copy").
+        level : int, optional
+            The logging level to use for the log message. Defaults to
+            `logging.DEBUG`.
+        """
+        logging.log(
+            level,
+            "BQ %s: job_id=%s location=%s state=%s bytes_processed=%s bytes_billed=%s slot_millis=%s "
+            "dml_rows=%s reference_tables=%s",
+            label,
+            job.job_id,
+            job.location,
+            job.state,
+            getattr(job, "total_bytes_processed", None),
+            getattr(job, "total_bytes_billed", None),
+            getattr(job, "slot_millis", None),
+            getattr(job, "num_dml_affected_rows", None),
+            getattr(job, "referenced_tables", None),
+        )
+
+    def run_job(
+        self, label: str, sql: str, job_config: bigquery.QueryJobConfig | None = None
+    ) -> bigquery.job.QueryJob:
+        """Run a BigQuery job with the given SQL and configuration.
+
+        Parameters
+        ----------
+        label : str
+            A label for the job, typically indicating the type of operation
+            (e.g., "insert", "delete", "copy").
+        sql : str
+            The SQL query to execute.
+        job_config : bigquery.QueryJobConfig, optional
+            Configuration for the job, such as query parameters or write
+            dispositions. If not provided, a default configuration will be
+            used.
+
+        Returns
+        -------
+        bigquery.job.QueryJob
+            The BigQuery job object representing the executed query. This can
+            be used to check the status of the job, retrieve results, or log
+            additional details.
+        """
+        job = self._bq_client.query(sql, job_config=job_config, location=self.dataset.location)
+        job.result()  # Wait for the job to complete
+        self.log_job(job, label)
+        return job
+
+
+class NoPromotableChunksError(Exception):
+    """Exception raised when there are no promotable chunks available."""
+
+    pass
+
+
+AnyBQJob: TypeAlias = (
+    bigquery.job.QueryJob
+    | bigquery.job.LoadJob
+    | bigquery.job.CopyJob
+    | bigquery.job.ExtractJob
+    | bigquery.job.UnknownJob
+)
+
+
+class ReplicaChunkPromoter:
+    """Class to promote replica chunks in BigQuery.
+
+    Parameters
+    ----------
+    runner : QueryRunner, optional
```

Comment:
```suggestion
    runner : `QueryRunner`, optional
```

Please use backticks consistently everywhere.

Reply:
I've fixed this throughout the codebase and will try to be consistent about it in the future.